### PR TITLE
Stabilize the Duration API

### PR DESCRIPTION
**Pull request by sfackler**
This commit stabilizes the `std::time` module and the `Duration` type.
`Duration::span` remains unstable, and the `Display` implementation for
`Duration` has been removed as it is still being reworked and all trait
implementations for stable types are de facto stable.

This is a [breaking-change] to those using `Duration`'s `Display`
implementation.

I'm opening this PR as a platform for discussion - there may be some method renaming to do as part of the stabilization process.

### CODE REVIEW

**Comment by sfackler** [comment]
Internals discussion: https://internals.rust-lang.org/t/final-comment-period-for-the-duration-api/2325

File: src/librustc/metadata/loader.rs [hunk 1]
```diff
@@ -721,7 +721,7 @@ fn get_metadata_section(target: &Target, filename: &Path)
     let dur = Duration::span(|| {
         ret = Some(get_metadata_section_imp(target, filename));
     });
-    info!("reading {:?} => {}", filename.file_name().unwrap(), dur);
```
Comment (nagisa) [inline src/librustc/metadata/loader.rs]: why?

File: src/libstd/time/duration.rs [hunk 2]
```diff
@@ -166,20 +167,6 @@ impl Div<u32> for Duration {
     }
 }
 
-impl fmt::Display for Duration {
```
Comment (nagisa) [inline src/libstd/time/duration.rs]: Why?

**Comment by alexcrichton** [comment]
I'm happy with this course of action. Could you also tag the trait implementations with the same stability tag as well?

**Comment by aturon** [comment]
I'm likewise fine with this general direction, modulo the renamings discussion on the internals thread. Will try to weigh in there soon.

**Comment by brson** [comment]
I also think it's time to get this done, and don't have opinions about the fine details - seems like this type has been well-polished at this point.

I'd definitely like to hear @lifthrasiir's opinions about this since [rust-chrono](https://github.com/lifthrasiir/rust-chrono) depends on it.

**Comment by lifthrasiir** [comment]
I have no particular objection, as Chrono has been a (relatively) happy user of the in-tree `Duration` type. I would have to adjust Chrono to deal with the removal of `Display` implementation but I can deal with it.

**Comment by sfackler** [comment]
I've updated the PR to rename `secs` to `as_secs` and `extra_nanos` to `subsec_nanos` in response to feedback from the internals discussion.

**Comment by aturon** [comment]
Can you keep deprecated versions of the old method names? We do this even for unstable APIs to ease changes in.

Otherwise, I'm ready to r+ assuming everyone is on board backporting this stabilization to 1.3 (which has essentially no downside, AFAICT.)

**Comment by steveklabnik** [comment]
>  (which has essentially no downside, AFAICT.)

The downside is a deviation from our standard release process. Yes, it's a bit painful to just miss a train, but at the same time, I don't want us to get in the habit of it.

**Comment by sfackler** [comment]
I'm assuming I should leave the old versions unstable and then cut them out in a release or two?

**Comment by sfackler** [comment]
Updated

**Comment by sfackler** [comment]
Should I change the feature for the stable functionality? Tidy's complaining:
```
error: feature 'duration' is both stable and unstable
```

**Comment by alexcrichton** [comment]
Yeah it's fine to just change the feature name of the deprecated methods to something new like `duration_old`.

**Comment by aturon** [comment]
FYI: we're going to discuss the backporting question in the core team meeting today.

**Comment by sfackler** [comment]
Cool. Adjusted the feature for the old methods.

**Comment by aturon** [comment]
We discussed this at the [core team meeting](https://github.com/rust-lang/meeting-minutes/blob/master/core-team/meeting-2015-08-10.md#duration-stabilization) and the consensus is that it's fine to backport this stabilization to 1.3 beta. In the future, we hope to avoid this by having more explicit "go/no go" subteam meetings around stabilization, sufficiently prior to the beta being cut.

**Comment by sfackler** [comment]
@bors: r-

(need to fix some benchmarks)

**Pull request merged**